src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -1248,19 +1249,24 @@ type IncrementalBuilder(initialState: IncrementalBuilderInitialState, state: Inc
             let t2 = MaxTimeStampInDependencies state.logicalStampedFileNames fileSlot
             max t1 t2
 
-    let gate = obj()
+    let semaphore = new SemaphoreSlim(1,1)
+
     let mutable currentState = state 
 
     let setCurrentState state cache (ct: CancellationToken) =
-        lock gate (fun () ->
-            ct.ThrowIfCancellationRequested()
-            currentState <- computeStampedFileNames initialState state cache
-        )
+        node {
+            do! semaphore.WaitAsync(ct) |> NodeCode.AwaitTask
```

Comment:
Since we did have examples where something in FCS got stuck, what about using explicitly defined (and overly big) **timeouts** for operations like this, to turn hanging into a visible error?
**WaitAsync** supports it.

(for dogfood version only)

Reply:
For dogfooding, wouldn't it be better to leave it, to catch locsk easier? I.e. I can dogfood it in roslyndev, and attach to it when I see a problem?